File: frontend/src/types/marketData.ts
export type MarketBar = {
  symbol: string;
  timeframe: string;
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  vwap?: number | null;
  trades_count?: number | null;
};

export type StreamPreview = {
  symbol: string;
  timeframe: string;
  source: string;
  bars: MarketBar[];
};

export type TrackedSymbol = {
  ticker: string;
  name: string | null;
  exchange: string | null;
  asset_type: string;
  is_active: boolean;
  market?: string | null;
  primary_exchange?: string | null;
  security_type?: string | null;
  currency?: string | null;
  round_lot_size?: number | null;
  minimum_order_size?: number | null;
  last_updated_utc?: string | null;
  last_price?: number | null;
  change?: number | null;
  change_percent?: number | null;
  quote_timestamp?: string | null;
  quote_source?: string | null;
};

export type MarketUniversePage = {
  items: TrackedSymbol[];
  next_cursor: string | null;
  source: string;
  as_of: string;
};

File: backend/app/schemas/market_data.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BarResponse(BaseModel):
    symbol: str
    timeframe: str
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: int
    vwap: float | None = None
    trades_count: int | None = None


class BackfillRequest(BaseModel):
    symbol: str
    timeframe: str = "1min"
    start: datetime
    end: datetime
    source: str = Field(default="auto", description="auto, eodhd, ibkr, polygon, finnhub, or synthetic")


class BackfillResponse(BaseModel):
    symbol: str
    timeframe: str
    inserted: int
    source: str
    start: datetime
    end: datetime


class StreamPreviewResponse(BaseModel):
    symbol: str
    timeframe: str
    source: str
    bars: list[BarResponse]


class MarketStatusResponse(BaseModel):
    exchange: str
    holiday: str | None = None
    is_open: bool
    session: str
    timezone: str
    timestamp: datetime
    provider: str


class SymbolLookupResponse(BaseModel):
    ticker: str
    name: str | None = None
    exchange: str | None = None
    asset_type: str
    is_active: bool
    market: str | None = None
    primary_exchange: str | None = None
    security_type: str | None = None
    currency: str | None = None
    round_lot_size: int | None = None
    minimum_order_size: int | None = None
    last_updated_utc: datetime | None = None
    last_price: float | None = None
    change: float | None = None
    change_percent: float | None = None
    quote_timestamp: datetime | None = None
    quote_source: str | None = None


class MarketUniverseResponse(BaseModel):
    items: list[SymbolLookupResponse]
    next_cursor: str | None = None
    source: str
    as_of: datetime
